---
title: "An overview of the epistack package"
author: "Guillaume Devailly"
package: epistack
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Using epistack}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

The `epistack` package main objective is the visualizations of stacks 
of genomic tracks (such as, but not restricted to, ChIP-seq or
DNA methyation data)
centered at genomic regions of interest. `epistack` needs three 
different inputs:

- a genomic score objects, such as ChIP-seq coverage or DNA methylation values, 
provided as a `GRanges` (easily obtained from `bigwig` or `bam` files)
- a list of feature of interest, such as peaks or transcription start sites,
provided as a `GRanges` (easily obtained from `gtf` or `bed` files)
- a score to sort the features, such as peak height of gene expression value

The following document will present two use cases:
- plotting ChIP-seq signal at peaks to investigate peak quality
- plotting ChIP-seq signal at gene promoters sorted according to gene expression
to investigate the associations between an epigenetic mark and gene expression

Tips and other `epistack` functionalities will then be presented.

# example data

To illustrate `epistack` fucntionalities, we will use example ChIP-seq data
used in the 
[2016 CASME course](https://www.bioconductor.org/help/course-materials/2016/CSAMA/lab-5-chipseq/Epigenetics.html)
"Basics of ChIP-seq data analysis" 
by Aleksandra Pekowska and Simon Anders.
Data can be found in this github repository: 
[https://github.com/Bioconductor/CSAMA2016/tree/master/lab-5-chipseq/EpigeneticsCSAMA/inst/bedfiles](github.com/Bioconductor/CSAMA2016/tree/master/lab-5-chipseq/EpigeneticsCSAMA/inst/bedfiles)

There is now need to download the data as it can be directly parsed remotely.
The data consists of 2 H3K27mac ChIP-seq replicates, an input control,
and one list of peak for each replicates. It has been generated in 
mouse Embryonic Stem cells and been subseted to have only data from chromosome 6
to allow fast vignette generation (but `epistack` can deal with whole genome 
ChIP-seq data!).

```{r examplepath}
path_reads <- c(
    read1 = "https://raw.githubusercontent.com/Bioconductor/CSAMA2016/master/lab-5-chipseq/EpigeneticsCSAMA/inst/bedfiles/H3K27ac_rep1_filtered_ucsc_chr6.bed",
    read2 = "https://raw.githubusercontent.com/Bioconductor/CSAMA2016/master/lab-5-chipseq/EpigeneticsCSAMA/inst/bedfiles/H3K27ac_rep2_filtered_ucsc_chr6.bed",
    input = "https://raw.githubusercontent.com/Bioconductor/CSAMA2016/master/lab-5-chipseq/EpigeneticsCSAMA/inst/bedfiles/ES_input_filtered_ucsc_chr6.bed"
)

path_peaks <- c(
    peak1 = "https://raw.githubusercontent.com/Bioconductor/CSAMA2016/master/lab-5-chipseq/EpigeneticsCSAMA/inst/bedfiles/Rep1_peaks_ucsc_chr6.bed",
    peak2 = "https://raw.githubusercontent.com/Bioconductor/CSAMA2016/master/lab-5-chipseq/EpigeneticsCSAMA/inst/bedfiles/Rep2_peaks_ucsc_chr6.bed"
)
```

# Example 1: ChIP-seq coverage at peaks

## data loading

We first need to reads the peaks using `r Biocpkg("rtracklayer")`:
```{r peak_loading, message=FALSE}
library(rtracklayer)
peaks <- lapply(path_peaks, import)
```

Peaks from each replicates can be merged using `r Biocpkg("GenomicRanges")`
`union()` function (loaded with `r Biocpkg("rtracklayer")`). 
We then rescue peaks metadata, and compute a new `mean_score` that we use to arrange our peak list.

```{r peak_merge, message=FALSE}
merged_peaks <- union(peaks[[1]], peaks[[2]])

scores_rep1 <- double(length(merged_peaks))
scores_rep1[findOverlaps(peaks[[1]], merged_peaks, select = "first")] <- peaks[[1]]$score

scores_rep2 <- double(length(merged_peaks))
scores_rep2[findOverlaps(peaks[[2]], merged_peaks, select = "first")] <- peaks[[2]]$score

peak_type <- ifelse(
    scores_rep1 != 0 & scores_rep2 != 0, "Both", ifelse(
        scores_rep1 != 0, "Rep1 only", "Rep2 only"
    )
)

mcols(merged_peaks) <- DataFrame(scores_rep1, scores_rep2, peak_type)
merged_peaks$mean_scores <- apply((mcols(merged_peaks)[, c("scores_rep1", "scores_rep2")]), 1, mean)
merged_peaks <- merged_peaks[order(merged_peaks$mean_scores, decreasing = FALSE), ]
rm(scores_rep1, scores_rep2, peak_type)

merged_peaks
```

We then import the ChIP-seq reads. In the example datasets,
they are provided as .bed files, but for ChIP-seq we recommand
importing .bigwig coverage files. Bam files can also be imported using 
`GenomicAlignments::readGAlignment*()`.

```{r read_loading}
reads <- lapply(path_reads, import)
```

## Generating coverage matrices

We now need to generate coverage matrices with ChIP-seq coverage signal
summarize around each ChIP-seq peaks.
Several tools exists to generate such coverage matrix. We will demonstrate the
`normalizeToMatrix()` method from `r Biocpkg("EnrichedHeatmap")`.
Other alternatives include `r Biocpkg("Repitools")`' `featureScores()`,
or `r Biocpkg("seqplots")`' `getPlotSetArray()`.

We will focus on the regions around peak centers, extended from +-5kb with
a window size of 250 bp. We keep track of the extent of our region of interest
in a `xlab` variable, and specify unique column names for each matrix.

When using ChIP-seq data from a bigwig file, use the `value_column` parameter
of the `normalizeToMatrix()` function.

```{r coverage_matrices, message=FALSE}
summary(width(merged_peaks))

library(EnrichedHeatmap)

coverage_matrices <- lapply(
    reads,
    function(x) {
        normalizeToMatrix(
            x,
            resize(merged_peaks, width = 1, fix = "center"),
            extend = 5000, w = 250, 
            mean_mode = "coverage"
        )
    }
)

colnames(coverage_matrices[[1]]) <- paste("Rep1" , seq_len(ncol(coverage_matrices[[1]])), sep = "_")
colnames(coverage_matrices[[2]]) <- paste("Rep2" , seq_len(ncol(coverage_matrices[[2]])), sep = "_")
colnames(coverage_matrices[[3]]) <- paste("Input", seq_len(ncol(coverage_matrices[[3]])), sep = "_")

xlabs <- c("-5kb", "peak center", "+5kb")
```

We then add each matrix to the `merged_peaks` object, and it should be ready for plotting.

```{r ready_to_plot}
mcols(merged_peaks) <- cbind(
    mcols(merged_peaks), 
    coverage_matrices[[1]], 
    coverage_matrices[[2]],
    coverage_matrices[[3]]
)
```

## Plotting

The `plotEpistack()` function will use the `merged_peaks` object to generate a
complex representation of the ChIP-seq signals around
the genomic feature of interests (here ChIP-seq peaks).
It has quite an important number of parameters.

```{r peak_plot, fig.wide = TRUE}
library(epistack)

plotEpistack(
    merged_peaks,
    patterns = c("^Rep1_", "^Rep2_", "^Input_"),
    tints = c("dodgerblue", "firebrick1", "grey"), 
    titles = c("Rep1", "Rep2" , "Input"),
    x_labels = xlabs,
    zlim = c(0, 4), ylim = c(0, 4), 
    metric_col = "mean_scores", metric_title = "Peak score", metric_label = "log10(score)",
    metric_transfunc = function(x) log10(x),
    npix_height = 300
)
```

If a `bin` column is present in the epistack GRanges object, than it will 
be used to annotate the figure and to generate one average profile per bin
in the lower panels. Here we use the `peak_type` as our bin column.

```{r peak_plot_bin, fig.wide = TRUE}

merged_peaks$bin <- merged_peaks$peak_type

plotEpistack(
    merged_peaks,
    patterns = c("^Rep1_", "^Rep2_", "^Input_"),
    tints = c("dodgerblue", "firebrick1", "grey"), 
    titles = c("Rep1", "Rep2" , "Input"),
    x_labels = xlabs,
    zlim = c(0, 4), ylim = c(0, 4), 
    metric_col = "mean_scores", metric_title = "Peak score", metric_label = "log10(score)",
    metric_transfunc = function(x) log10(x),
    bin_palette = colorRampPalette(c("darkorchid1", "dodgerblue", "firebrick1"))
)

```

# Example 2: ChIP-seq coverage at gene promoters sorted according to
expression levels

# Obtaining the TSS coordinates

We now want to plot the ChIP-seq signals at gene promoters, or more
precisely around gene Transcription Start Sites (TSS).
TSS coordinate can be obtained from various sources. Here we will access
the [Ensembl](https://www.ensembl.org/index.html) annotations using 
`r Biocpkg("biomaRt")`,
, but one could
also download a `.gtf` file and parse it using `r Biocpkg("rtracklayer")`'s
`import()`, or use `r Biocpkg("AnnotationHub")` and 
`r Biocpkg("ensembldb")`. It is however important to work with the same genome
version has the one used to align the ChIP-seq reads (here the mm9/NCBIm37
genome).

```{r accessing_ensembl}
library(biomaRt)

mart <- useMart(
    biomart = "ENSEMBL_MART_ENSEMBL", 
    dataset = "mmusculus_gene_ensembl", 
    host = "may2012.archive.ensembl.org"
)
ds <- useDataset('mmusculus_gene_ensembl', mart = mart)
chr6_genes <- getBM(
    attributes = c(
        'ensembl_gene_id','external_gene_id',
        'chromosome_name','start_position',
        'end_position','strand'), 
    filters = 'chromosome_name', values = 6, # only chr6 in this example
    mart = ds
)

chr6_genes <- with(chr6_genes, GRanges(
    seqnames = paste0("chr", chromosome_name),
    IRanges(start_position, end_position),
    strand = strand,
    gene_id = ensembl_gene_id,
    gene_symbol = external_gene_id
))

chr6_genes
```




As for the previous example, we can now use this 

## expression data --------------
