---
title: "An overview of the epistack package"
author: "Guillaume Devailly"
package: epistack
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{using_epistack}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

The `epistack` package main objective is the visualizations of stacks 
of genomic tracks (such as, but not restricted to, ChIP-seq or
DNA methyation data)
centered at genomic regions of interest. `epistack` needs three 
different inputs:

- a genomic score objects, such as ChIP-seq coverage or DNA methylation values, 
provided as a `GRanges` (easily obtained from `bigwig` or `bam` files)
- a list of feature of interest, such as peaks or transcription start sites,
provided as a `GRanges` (easily obtained from `gtf` or `bed` files)
- a score to sort the features, such as peak height of gene expression value

The following document will present two use cases:
- plotting ChIP-seq signal at peaks to investigate peak quality
- plotting ChIP-seq signal at gene promoters sorted according to gene expression
to investigate the associations between an epigenetic mark and gene expression

Tips and other `epistack` functionalities will then be presented.

```{r setup}
library(epistack)
```
